util/src/main/java/tc/oc/pgm/util/bukkit/Platform.java

```diff
@@ -2,62 +2,103 @@
+import tc.oc.pgm.util.nms.v1_10_12.NMSHacks1_10_12;
+import tc.oc.pgm.util.nms.v1_8.NMSHacks1_8;
 import tc.oc.pgm.util.nms.v1_8.NMSHacksSportPaper;
 import tc.oc.pgm.util.nms.v1_9.NMSHacks1_9;
-import tc.oc.pgm.util.reflect.ReflectionUtils;
 
 public enum Platform {
   UNKNOWN("UNKNOWN", "UNKNOWN", NMSHacksNoOp.class, false),
   SPORTPAPER_1_8("SportPaper", "1.8", NMSHacksSportPaper.class, false),
-  SPIGOT_1_8(
-      "Spigot",
-      "1.8", // NMSHacks1_8 causes issues with other versions, get it dynamically
-      (Class<? extends NMSHacksPlatform>)
-          ReflectionUtils.getClassFromName("tc.oc.pgm.util.nms.v1_8.NMSHacks1_8"),
-      false),
-  PAPER_1_8(
-      "Paper",
-      "1.8", // NMSHacks1_8 causes issues with other versions, get it dynamically
-      (Class<? extends NMSHacksPlatform>)
-          ReflectionUtils.getClassFromName("tc.oc.pgm.util.nms.v1_8.NMSHacks1_8"),
-      false),
+  SPIGOT_1_8("Spigot", "1.8", () -> NMSHacks1_8.class, false),
+  PAPER_1_8("Paper", "1.8", () -> NMSHacks1_8.class, false),
   SPIGOT_1_9("Spigot", "1.9", NMSHacks1_9.class, true),
-  PAPER_1_9("Paper", "1.9", NMSHacks1_9.class, true);
+  PAPER_1_9("Paper", "1.9", NMSHacks1_9.class, true),
+  SPIGOT_1_10_12("Spigot", "1.10", "1.12", NMSHacks1_10_12.class, true),
+  PAPER_1_10_12("Paper", "1.10", "1.12", NMSHacks1_10_12.class, true);
 
   private static ClassLogger logger = ClassLogger.get(Platform.class);;
   public static Platform SERVER_PLATFORM = computeServerPlatform();
 
   private static Platform computeServerPlatform() {
     String versionString = Bukkit.getServer().getVersion();
     for (Platform platform : Platform.values()) {
-      if (versionString.contains(platform.variant)
-          && versionString.contains("MC: " + platform.majorVersion)) {
-        return platform;
+      for (String supportedMajorVersion : platform.getSupportedMajorVersions()) {
+        if (versionString.contains(platform.variant)
+            && versionString.contains("MC: " + supportedMajorVersion)) {
+          return platform;
+        }
       }
     }
     return UNKNOWN;
   }
 
   private final String variant;
-  private final String majorVersion;
-  private final Class<? extends NMSHacksPlatform> nmsHacksClass;
+  private final String majorVersionFirst;
+  private final String majorVersionLast;
+  private final Supplier<Class<? extends NMSHacksPlatform>> nmsHacksClassSupplier;
   private final boolean requiresProtocolLib;
 
   Platform(
       String variant,
       String majorVersion,
+      Supplier<Class<? extends NMSHacksPlatform>> nmsHacksClassSupplier,
+      boolean requiresProtocolLib) {
+    this.variant = variant;
+    this.majorVersionFirst = majorVersion;
+    this.majorVersionLast = majorVersion;
+    this.nmsHacksClassSupplier = nmsHacksClassSupplier;
+    this.requiresProtocolLib = requiresProtocolLib;
+  }
+
+  Platform(
+      String variant,
+      String majorVersion,
+      Class<? extends NMSHacksPlatform> nmsHacksClass,
```

Comment:
if you keep on adding constructors for each possible different combination, this is going to explode quickly, consider just having supplier-based constructors, make all of them pass in the supplier, and maybe even force major & minor version, and just pass in the same version

Reply:
Yeah I can simplify this